### PR TITLE
feat(prover)!: Add EIP4844 support for provers subsystem

### PR DESCRIPTION
**Pull request by EmilLuta**
This PR adds support for EIP4844 as part of 1.4.2 release.

To support EIP4844, the prover needs to:
- prove EIP4844 blobs
- inject the result into scheduler proof

EIP4844 is special cased for 1.4.2. Later (1.5.0) it will behave like any other circuit.

How EIP4844 works?

Blob proofs are generated at basic/ aggregation_round 0 stage and are marked as final nodes to be inserted into scheduler. If there is a single blob, the [witness for the second blob needs to be zeroed out](https://github.com/matter-labs/era-zkevm_circuits/blob/v1.4.1/src/scheduler/mod.rs#L1149) whilst the [proof still needs to be injected](https://github.com/matter-labs/era-zkevm_circuits/blob/v1.4.1/src/scheduler/mod.rs#L1165) (any proof will work). Blobs are stored in `witness_inputs_fri` and `scheduler_dependency_tracker` is aware of the proofs as `circuit_255_final_prover_job_id_0` for the first blob and `circuit_255_final_prover_job_id_1` for the second blob. (circuit id 255 is blob, until the code-base migrates to 1.5.0, when it will become a regular circuit).

### CODE REVIEW

File: prover/prover_fri/src/utils.rs [hunk 1]
```diff
@@ -118,12 +119,18 @@ pub async fn save_proof(
             .await;
     }
     if job_metadata.is_node_final_proof {
+        let circuit_id = if job_metadata.circuit_id == EIP_4844_CIRCUIT_ID {
+            EIP_4844_CIRCUIT_ID
+        } else {
+            get_base_layer_circuit_id_for_recursive_layer(job_metadata.circuit_id)
+        };
```
Comment (RomanBrodetski) [inline prover/prover_fri/src/utils.rs:126]: I don't love this but given that it'll be gone with 1.5.0 I think it's the right compromise

File: prover/witness_generator/src/basic_circuits.rs [hunk 2]
```diff
@@ -661,17 +703,50 @@ async fn generate_witness(
...
     };
 
+    let mut eip_4844_blobs = eip_4844_blobs.blobs;
+
+    let mut eip_4844_circuits = vec![];
+    let mut eip_4844_witnesses = vec![];
+
+    let single_blob = eip_4844_blobs.len() == 1;
+    if single_blob {
+        eip_4844_blobs.push(eip_4844_blobs[0].clone());
```
Comment (RomanBrodetski) [inline prover/witness_generator/src/basic_circuits.rs:719]: wait, what? If we only have one blob, we clone it and make it appear that it's two blobs? If so let's at least add a comment

File: prover/witness_generator/src/basic_circuits.rs [hunk 3]
```diff
@@ -661,17 +703,50 @@ async fn generate_witness(
...
+    }
+    for blob in eip_4844_blobs {
+        let (eip_4844_circuit, eip_4844_witness) = generate_eip4844_circuit_and_witness(
+            blob.to_vec(),
+            "/home/evl/zksync_era/trusted_setup.json",
+        );
+        eip_4844_circuits.push(eip_4844_circuit);
+        eip_4844_witnesses.push(eip_4844_witness.closed_form_input.observable_output);
+    }
```
Comment (RomanBrodetski) [inline prover/witness_generator/src/basic_circuits.rs:728]: I prefer a declarative way: would something like `let (eip_4844_circuits, eip_4844_witnesses) = eip_4844_blobs.iter().map(generate_eip4844_circuit_and_witness).unzip()` work? It's a little hard to keep track on all the mutable lists here -when and why they are populated

**Review by RomanBrodetski — COMMENTED** [review]
Overall looks good, left some stylistic comments.

File: etc/env/base/contracts.toml [hunk 4]
```diff
@@ -40,7 +40,7 @@ L2_WETH_TOKEN_IMPL_ADDR="0x5E6D086F5eC079ADFF4FB3774CDf3e8D6a34F7E9"
 L2_WETH_TOKEN_PROXY_ADDR="0x5E6D086F5eC079ADFF4FB3774CDf3e8D6a34F7E9"
 FRI_RECURSION_LEAF_LEVEL_VK_HASH ="0x062362cb3eaf1f631406cbe19bf2a2c5d0d9ea69d069309a6003addae9f387be"
 FRI_RECURSION_NODE_LEVEL_VK_HASH ="0x5a3ef282b21e12fe1f4438e5bb158fc5060b160559c5158c6389d62d9fe3d080"
-FRI_RECURSION_SCHEDULER_LEVEL_VK_HASH ="0xf12256f7ba1ab223ba82abb19c8366c6677596897f9e5a82016977113e52ed5d"
+FRI_RECURSION_SCHEDULER_LEVEL_VK_HASH ="0x1405880dc3317d635bddb0ab62bf5d013e5d1f462161c1f7ac3289c7fef956da"
```
[the remark this answers is not in the document]
Comment (mm-zk) [reply etc/env/base/contracts.toml:43]: then you should also update the SNARK_WRAPPER_VK_HASH

File: etc/env/base/fri_prover.toml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 [fri_prover]
-setup_data_path="/usr/src/setup-data"
+setup_data_path="vk_setup_data_generator_server_fri/data/"
```
[the remark this answers is not in the document]
Comment (mm-zk) [reply etc/env/base/fri_prover.toml:2]: ack - normally I'd do such changes in separate PR, as they are not related to the 4844. But I'll leave that decision to you.

File: core/lib/prover_interface/src/api.rs [hunk 6]
```diff
@@ -9,12 +9,80 @@ use zksync_types::{
...
+pub struct Eip4844Blobs {
+    blobs: Vec<Blob>,
+}
+
+impl Eip4844Blobs {
+    pub fn blobs(self) -> Vec<Blob> {
+        self.blobs
+    }
+
+    fn get_rounded_blob_len(blobs_len: usize) -> usize {
```
Comment (mm-zk) [inline core/lib/prover_interface/src/api.rs:31]: comment

File: core/lib/dal/src/fri_witness_generator_dal.rs [hunk 7]
```diff
@@ -39,6 +39,7 @@ impl FriWitnessGeneratorDal<'_, '_> {
         block_number: L1BatchNumber,
         object_key: &str,
         protocol_version_id: FriProtocolVersionId,
+        eip_4844_blobs: Vec<u8>,
```
Comment (mm-zk) [inline core/lib/dal/src/fri_witness_generator_dal.rs:42]: nit: maybe pass as reference (as you're using them as reference anyways)

File: core/lib/dal/src/fri_witness_generator_dal.rs [hunk 8]
```diff
@@ -69,7 +72,7 @@ impl FriWitnessGeneratorDal<'_, '_> {
         last_l1_batch_to_process: u32,
```
Comment (mm-zk) [inline core/lib/dal/src/fri_witness_generator_dal.rs:72]: please add comment to the method, and explain what it returns - especially the Vec<u8> part (is it raw pubdata ? ome blob ? empty)

File: core/lib/dal/src/fri_scheduler_dependency_tracker_dal.rs [hunk 9]
```diff
@@ -69,15 +71,27 @@ impl FriSchedulerDependencyTrackerDal<'_, '_> {
         circuit_id: u8,
```
Comment (mm-zk) [inline core/lib/dal/src/fri_scheduler_dependency_tracker_dal.rs:71]: add comment and explain what "blob_ordering" means

File: core/lib/dal/src/fri_prover_dal.rs [hunk 10]
```diff
@@ -249,14 +253,15 @@ impl FriProverDal<'_, '_> {
         for (sequence_number, (circuit_id, circuit_blob_url)) in
             circuit_ids_and_urls.iter().enumerate()
         {
+            let is_final = *circuit_id == types::EIP_4844_CIRCUIT_ID;
```
Comment (mm-zk) [inline core/lib/dal/src/fri_prover_dal.rs:256]: comment why we treat this special

File: prover/prover_fri/src/utils.rs [hunk 11]
```diff
@@ -118,12 +119,18 @@ pub async fn save_proof(
...
         transaction
             .fri_scheduler_dependency_tracker_dal()
             .set_final_prover_job_id_for_l1_batch(
-                get_base_layer_circuit_id_for_recursive_layer(job_metadata.circuit_id),
+                circuit_id,
                 job_id,
                 job_metadata.block_number,
+                job_metadata.sequence_number,
```
Comment (mm-zk) [inline prover/prover_fri/src/utils.rs:133]: comment why we're passing sequence_number as blob_ordering

File: prover/witness_generator/src/scheduler.rs [hunk 12]
```diff
@@ -87,13 +93,21 @@ impl SchedulerWitnessGenerator {
             _marker: std::marker::PhantomData,
         };
 
+        let eip_4844_config = ProofConfig {
```
Comment (mm-zk) [inline prover/witness_generator/src/scheduler.rs:96]: please use pub fn eip4844_proof_config() from test_harness:
https://github.com/matter-labs/era-zkevm_test_harness/blob/b9b969516f83c8a66ae1d71dd30ee42553b8fa25/circuit_definitions/src/lib.rs#L49

File: core/lib/dal/migrations/20240226120530_add_support_for_eip4844.up.sql [hunk 13]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE witness_inputs_fri ADD COLUMN eip_4844_blobs BYTEA;
+
+ALTER TABLE scheduler_dependency_tracker_fri ADD COLUMN IF NOT EXISTS eip_4844_final_prover_job_id_0 BIGINT DEFAULT NULL;
+ALTER TABLE scheduler_dependency_tracker_fri ADD COLUMN IF NOT EXISTS eip_4844_final_prover_job_id_1 BIGINT DEFAULT NULL;
```
Comment (perekopskiy) [inline core/lib/dal/migrations/20240226120530_add_support_for_eip4844.up.sql:4]: nit (optional): `DEFAULT NULL` can be omitted

File: core/lib/dal/migrations/20240226120530_add_support_for_eip4844.up.sql [hunk 14]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE witness_inputs_fri ADD COLUMN eip_4844_blobs BYTEA;
```
Comment (perekopskiy) [inline core/lib/dal/migrations/20240226120530_add_support_for_eip4844.up.sql:1]: nit: `IF NOT EXISTS`

File: core/lib/dal/src/fri_witness_generator_dal.rs [hunk 15]
```diff
@@ -47,29 +50,34 @@ impl FriWitnessGeneratorDal<'_, '_> {
...
                 )
             VALUES
-                ($1, $2, $3, 'queued', NOW(), NOW())
+                ($1, $2, $3, $4, 'queued', NOW(), NOW())
             ON CONFLICT (l1_batch_number) DO NOTHING
             "#,
             block_number.0 as i64,
             object_key,
             protocol_version_id as i32,
+            bincode::serialize(&Vec::<u8>::from(eip_4844_blobs))
```
Comment (perekopskiy) [inline core/lib/dal/src/fri_witness_generator_dal.rs:65]: Why serialize `Vec<u8>`? It's already serialized

File: core/lib/dal/src/fri_witness_generator_dal.rs [hunk 16]
```diff
@@ -69,7 +72,7 @@ impl FriWitnessGeneratorDal<'_, '_> {
         last_l1_batch_to_process: u32,
         protocol_versions: &[FriProtocolVersionId],
         picked_by: &str,
-    ) -> Option<L1BatchNumber> {
+    ) -> Option<(L1BatchNumber, Vec<u8>)> {
```
[the remark this answers is not in the document]
Comment (RomanBrodetski) [reply core/lib/dal/src/fri_witness_generator_dal.rs]: all sounds good and yes BWIP will take care of this

File: core/lib/zksync_core/src/proof_data_handler/request_processor.rs [hunk 17]
```diff
@@ -118,13 +119,31 @@ impl RequestProcessor {
             }
         };
 
+        let storage_batch = self
```
Comment (perekopskiy) [inline core/lib/zksync_core/src/proof_data_handler/request_processor.rs:122]: nit(optional): you can add a separate method to load only `pubdata_input` as `l1_batch` has some other heavy columns

File: prover/prover_dal/migrations/20240226120310_add_support_for_eip4844.up.sql [hunk 18]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE witness_inputs_fri ADD COLUMN eip_4844_blobs BYTEA;
+
+ALTER TABLE scheduler_dependency_tracker_fri ADD COLUMN IF NOT EXISTS eip_4844_final_prover_job_id_0 BIGINT DEFAULT NULL;
+ALTER TABLE scheduler_dependency_tracker_fri ADD COLUMN IF NOT EXISTS eip_4844_final_prover_job_id_1 BIGINT DEFAULT NULL;
```
Comment (perekopskiy) [inline prover/prover_dal/migrations/20240226120310_add_support_for_eip4844.up.sql:4]: Same as above

**Review by perekopskiy — COMMENTED** [review]
Overall looks good! Left some minor comments

File: core/lib/dal/src/fri_witness_generator_dal.rs [hunk 19]
```diff
@@ -47,29 +50,34 @@ impl FriWitnessGeneratorDal<'_, '_> {
...
             protocol_version_id as i32,
+            bincode::serialize(&Vec::<u8>::from(eip_4844_blobs))
+                .expect("Failed to serialize 4844 blobs to bytes"),
         )
         .fetch_optional(self.storage.conn())
         .await
         .unwrap();
     }
 
+    /// Gets the next job to be executed. Returns the batch number and it's corresponding blobs.
```
Comment (RomanBrodetski) [inline core/lib/dal/src/fri_witness_generator_dal.rs:72]:
```suggestion
    /// Gets the next job to be executed. Returns the batch number and its corresponding blobs.
```

File: prover/witness_generator/src/scheduler.rs [hunk 20]
```diff
@@ -233,25 +247,41 @@ impl JobProcessor for SchedulerWitnessGenerator {
...
+    let mut eip_4844_proofs = vec![];
+    for wrapper in proofs {
+        match wrapper {
+            FriProofWrapper::Base(_) => anyhow::bail!(
+                "Expected only EIP4844 proofs for scheduler l1 batch {l1_batch_number}, got Base"
+            ),
+            FriProofWrapper::Recursive(_) =>
+                anyhow::bail!("Expected only EIP4844 proofs for scheduler l1 batch {l1_batch_number}, got Recursive"),
+            FriProofWrapper::Eip4844(eip_4844_proof) => {
+                eip_4844_proofs.push(eip_4844_proof);
```
[the remark this answers is not in the document]
Comment (RomanBrodetski) [reply prover/witness_generator/src/scheduler.rs:278]: would [this](https://stackoverflow.com/a/26370894) work?

**Review by RomanBrodetski — COMMENTED** [review]
good work!

**Review by perekopskiy — APPROVED** [review]

**Review by Deniallugo — DISMISSED** [review]

**Review by montekki — APPROVED** [review]

**Pull request merged**